app/controllers/uploads_controller.rb

```diff
@@ -270,6 +296,174 @@ def complete_external_upload
     end
   end
 
+  def create_multipart
+    RateLimiter.new(
+      current_user, "create-multipart-upload", CREATE_MULTIPART_RATE_LIMIT_PER_MINUTE, 1.minute
+    ).performed!
+
+    file_name = params.require(:file_name)
+    file_size = params.require(:file_size).to_i
+    upload_type = params.require(:upload_type)
+    content_type = MiniMime.lookup_by_filename(file_name)&.content_type
+
+    if file_size_too_big?(file_name, file_size)
+      return render_json_error(
+        I18n.t("upload.attachments.too_large", max_size_kb: SiteSetting.max_attachment_size_kb),
+        status: 422
+      )
+    end
+
+    begin
+      multipart_upload = Discourse.store.create_multipart(
+        file_name, content_type
+      )
+    rescue Aws::S3::Errors::ServiceError => err
+      debug_upload_error(err, "upload.create_mutlipart_failure")
+      return render_json_error(I18n.t("upload.failed"), status: 422)
+    end
+
+    upload_stub = ExternalUploadStub.create!(
+      key: multipart_upload[:key],
+      created_by: current_user,
+      original_filename: file_name,
+      upload_type: upload_type,
+      external_upload_identifier: multipart_upload[:upload_id],
+      multipart: true,
+      filesize: file_size
+    )
+
+    render json: {
+      external_upload_identifier: upload_stub.external_upload_identifier,
+      key: upload_stub.key,
+      unique_identifier: upload_stub.unique_identifier
+    }
+  end
+
+  def batch_presign_multipart_parts
+    part_numbers = params.require(:part_numbers)
+    unique_identifier = params.require(:unique_identifier)
+
+    RateLimiter.new(
+      current_user, "batch-presign", BATCH_PRESIGN_RATE_LIMIT_PER_MINUTE, 1.minute
+    ).performed!
+
+    part_numbers = part_numbers.map do |part_number|
+      validate_part_number(part_number)
+    end
+
+    external_upload_stub = ExternalUploadStub.find_by(
+      unique_identifier: unique_identifier, created_by: current_user
+    )
+    return render_404 if external_upload_stub.blank?
+
+    presigned_urls = {}
+    part_numbers.each do |part_number|
+      presigned_urls[part_number] = Discourse.store.presign_multipart_part(
+        upload_id: external_upload_stub.external_upload_identifier,
+        key: external_upload_stub.key,
+        part_number: part_number
+      )
+    end
+
+    if !multipart_upload_exists?(external_upload_stub)
+      return render_404
+    end
```

Comment:
```suggestion
    if !multipart_upload_exists?(external_upload_stub)
      return render_404
    end

    presigned_urls = {}
    part_numbers.each do |part_number|
      presigned_urls[part_number] = Discourse.store.presign_multipart_part(
        upload_id: external_upload_stub.external_upload_identifier,
        key: external_upload_stub.key,
        part_number: part_number
      )
    end
```

We should check if the `multipart_upload_exists?` for the stub first before generating `presigned_urls`.